### ISSUE TITLE
Problem with Thunderbolt 3

### ISSUE THREAD

**Issue by grandstudio01** [SchmockLord/Hackintosh-Intel-i9-10900k-AsRock-Z490-Phantom-ITX-TB3#14]
Follow These step and everything work fine but not Thunderbolt3.

Anyone can help mo figure it out why or what should i do to Fix it.

I use the ASROCK z490 phantom gaming itx/tb3

**Comment by phi-alek** [comment]
i have just done what was in the pictures for the bios settings and thundbolt works perfectly. even hotplugging.

**Comment by SchmockLord** [comment]
How about TB3 with the newest config?

**Comment by phi-alek** [comment]
Still works perfectly.

**Comment by webleon** [comment]
can't drive Thunderbolt3 either.
I followed every step in the README.md, using EFI in the latest release, iMac 20.2 config file, nothing changed but serials.

btw, the Zip file in the release missed some of the kexts. :)

**Comment by fidalgodev** [comment]
@SchmockLord in latest release I was having a random issue that happened once in a while, where the screen would go black, and the computer would crash. When I restarted I wouldn't get that error screen saying the computer crashed.

Could this be related to the missing kexts? Seems like a GPU related issue. I have the same exact GPU as you btw

**Comment by webleon** [comment]
@SchmockLord thank you for the quick response, can u also help with the thunderbolt3? it still doesn't work.
should I do extra work on the config.plist?

**Comment by SchmockLord** [comment]
Hey guys,

I have found this website here to generate SSDTs for Thunderbolt 3 Hotplug: https://hackindrom.zapto.org

Can you please test this file?

It is just the generated one from this website and then renamed to SSDT-TB3HP.aml

Just replace your one in OC/ACPI folder.

Best,
Chris
[SSDT-TB3HP.aml.zip](https://github.com/SchmockLord/Hackintosh-Intel-i9-10900k-AsRock-Z490-Phantom-ITX-TB3/files/5646918/SSDT-TB3HP.aml.zip)

**Comment by SchmockLord** [comment]
Now with the newest BIOS Beta 1.42, it is showing me TB3 at RP21 and not RP05 anymore.

So here is the same file again, but with RP21.

[SSDT-TB3HP.aml.zip](https://github.com/SchmockLord/Hackintosh-Intel-i9-10900k-AsRock-Z490-Phantom-ITX-TB3/files/5647014/SSDT-TB3HP.aml.zip)

**Comment by webleon** [comment]
> Now with the newest BIOS Beta 1.42, it is showing me TB3 at RP21 and not RP05 anymore.
> 
> So here is the same file again, but with RP21.
> 
> [SSDT-TB3HP.aml.zip](https://github.com/SchmockLord/Hackintosh-Intel-i9-10900k-AsRock-Z490-Phantom-ITX-TB3/files/5647014/SSDT-TB3HP.aml.zip)

Just tried this one, can see the basic info as a PCI device, but in the Thnderbolt section of sysinfo there are no drivers are loaded
<img width="1015" alt="Screen Shot 2020-12-07 at 21 40 53" src="https://user-images.githubusercontent.com/642111/101357990-34cd6e80-38d5-11eb-98d9-c65f0aab569e.png">
<img width="1015" alt="Screen Shot 2020-12-07 at 21 41 11" src="https://user-images.githubusercontent.com/642111/101357995-36973200-38d5-11eb-9d8e-bade588eec78.png">

**Comment by fidalgodev** [comment]
> > Now with the newest BIOS Beta 1.42, it is showing me TB3 at RP21 and not RP05 anymore.
> > So here is the same file again, but with RP21.
> > [SSDT-TB3HP.aml.zip](https://github.com/SchmockLord/Hackintosh-Intel-i9-10900k-AsRock-Z490-Phantom-ITX-TB3/files/5647014/SSDT-TB3HP.aml.zip)
> 
> Just tried this one, can see the basic info as a PCI device, but in the Thnderbolt section of sysinfo there are no drivers are loaded
> <img alt="Screen Shot 2020-12-07 at 21 40 53" width="1015" src="https://user-images.githubusercontent.com/642111/101357990-34cd6e80-38d5-11eb-98d9-c65f0aab569e.png">
> <img alt="Screen Shot 2020-12-07 at 21 41 11" width="1015" src="https://user-images.githubusercontent.com/642111/101357995-36973200-38d5-11eb-9d8e-bade588eec78.png">

You're not supposed to see the drivers under Thunderbolt as the implementation is different from the Apple one. You're only supposed to see it under PCI.

**Comment by SchmockLord** [comment]
@fidalgodev Do you have TB3 devices to test the functioanlity? I don't.

**Comment by fidalgodev** [comment]
> @fidalgodev Do you have TB3 devices to test the functioanlity? I don't.

Not here at the moment.
But I can confirm I have TB3 working and also Hotplug working. Im using 1.4 BIOS tho.

I'm using the SSDT-DTPG.aml and SSDT-Z490-TB3HP.aml from [this](https://www.tonymacx86.com/threads/gigabyte-z490-vision-d-thunderbolt-3-i5-10400-amd-rx-580.298642/) thread, it's a Gigabyte Z490 Vision D (has the exact same TB3 controller as this board, JHL7540 Titan Ridge.

Then i simply mapped the USB ports.

![photo](https://i.imgur.com/rQ9ik6C.png)

You can confirm TB3 and Hotplug is working if you find this entries using [IORegistryExplorer](https://github.com/khronokernel/IORegistryClone/blob/master/ioreg-302.zip)

![photo](https://www.tonymacx86.com/attachments/screenshot-2020-11-28-at-12-34-49-png.498785/)

**Comment by SchmockLord** [comment]
That is funny. Because the guy from the thread you have mentioned mostly copied my build. Nothing to say against it, but I can see that because he still has my typo in SSDT-DTPG.aml. Meanwhile, I have first renamed it to the correct name of the method DT**G**P to SSDT-DTGP.aml and later found out, that the method is already in the SSDT-Plug.aml. So I removed it.

Anyway. I have just tried out different variations of SSDTs to get Thunderbolt 3 Hot plug properly working. And some people still seem to have issues.

**Comment by fidalgodev** [comment]
> That is funny. Because the guy from the thread you have mentioned mostly copied my build. Nothing to say against it, but I can see that because he still has my typo in SSDT-DTPG.aml. Meanwhile, I have first renamed it to the correct name of the method DT**G**P to SSDT-DTGP.aml and later found out, that the method is already in the SSDT-Plug.aml. So I removed it.
> 
> Anyway. I have just tried out different variations of SSDTs to get Thunderbolt 3 Hot plug properly working. And some people still seem to have issues.

That I have no clue dude. But everything is working here.

One thing I don't understand still on this repo is the SSDT-Disable-CNVW.aml.

This is supposed to disable the mini pcie slot where the default WIFI card sits. The thing is, we don't want to disable it since we have a different card in there?

Also it doesn't even seem to be working because even with the SSDT my wifi card still works?

Correct me if im wrong please 😄

**Comment by SchmockLord** [comment]
I cannot recap where I got this SSDT-Disable-CNVW.aml from. I think someone mentioned it because I had sleep issues and found out my Mac is waking up by a CNVi/CNVW device.

And this SSDT doesn't disable the mini PCIe slot, it disables the CNVi/CNVW feature which is used by the Intel Wifi to share bandwidth with the CPU and RAM for Wifi. 

This is how I got the purpose of this SSDT.

**Comment by fidalgodev** [comment]
> I cannot recap where I got this SSDT-Disable-CNVW.aml from. I think someone mentioned it because I had sleep issues and found out my Mac is waking up by a CNVi/CNVW device.
> 
> And this SSDT doesn't disable the mini PCIe slot, it disables the CNVi/CNVW feature which is used by the Intel Wifi to share bandwidth with the CPU and RAM for Wifi.
> 
> This is how I got the purpose of this SSDT.

I see that makes sense! So technically we can disable CNVi?

**Comment by SchmockLord** [comment]
Yeah. After applying the SSDT, the CNVi feature didn't appear in the wake up triggers anymore.

**Comment by fidalgodev** [comment]
So does the new BETA 1.42 BIOS changes the port names for the thunderbolt port?

**Comment by fidalgodev** [comment]
@SchmockLord with your latest Thunderbolt SSDT don't you get a weird icon on the status bar? Like a card icon. When you click you have options to turn off the card or something.

I get this icon and I think this means the thunderbolt is not working?

**Comment by SchmockLord** [comment]
@fidalgodev Noticed the same, but it works flawless with my eGPU. Do you now a better SSDT-TB3.aml which you have tested to be working?

**Comment by fidalgodev** [comment]
The most annoying part is when you click on one of those sub menus the computer crashes...

I think the only thing failing on your SSDT is the controller name. If you noticed, under the Hackintool, the JHL7540 controller has the same XHC name as the normal USB controller. It should have XHC5 or something I think. So because it has the same as the normal USB controller, it thinks the thunderbolt controller is not being used? The port names on the thunderbolt controller are the same as some on the normal usb controller too.

I think [this ](https://drive.google.com/file/d/146HX16Fxsgu-fqjv10olfLUlQiQ0T23C/view?usp=sharing)one is working correctly.

**Comment by SchmockLord** [comment]
Yeah I think so too. That it needs to get a different name and that's all.

The one you sent me didn't work. Same as the one before: System crashes/restarts while macOS starts up.

Do you know how to edit the name it appears taking the current one from my repo?

But thanks.

**Comment by fidalgodev** [comment]
I think mine does not work for you because you're using bios 1.42 beta where TB3 controller is mapped to port RP21?

I'm using bios 1.4 (not beta) so it's on RP05 for me

**Comment by fidalgodev** [comment]
@SchmockLord can you disable all TB3 SSDT's, and then restart and use IORegistryExplorer and tell me what port does the TB3 controller appears on?

Mine is on RP21 even on 1.4 BIOS (not beta).
I assume you use 1.42 BETA but it's also on RP21?

Only on versions before 1.4 it was on RP05 I think

**Comment by fidalgodev** [comment]
Can you try [this](https://drive.google.com/file/d/1okA5JZ9V7b1mV1EVzYLo-jyjzr-2Z-Rl/view?usp=sharing) one?

This one is linked to the RP21 (correct TB3 port)

**Comment by fidalgodev** [comment]
The one I posted seems to be working for me.

![image](https://www.tonymacx86.com/attachments/screenshot-2020-12-14-at-16-02-34-png.500999)

When I plug a USBC to USB adapter, and then a usb 2.0 or 3.0 stick on it, the HS12 port is the one active (XHC).
Then I plug a USB C external drive into the usb c port, the SSP1 (XHC2) goes on.

Meaning normal USB is handled by the USB 3.1 controller and USB C devices by the Thunderbolt controller I guess.

**Comment by fidalgodev** [comment]
Im also very confused on how the SSDT that you have is working for you?

I just checked and it's mapping the RP05 as the Thunderbolt device...
While you said the Thunderbolt device appears under RP21 on IORegistryExplorer?

**Comment by SchmockLord** [comment]
@fidalgodev Sorry for the confusion. Must have been drunk when I uploaded this file. But at least this showed me, that any TB3 SSDT variation that is addressing the correct port RP21, is crashing my MacOS as soon as the Thunderbolt eGPU case is connected. No matter if I connect it before or after booting. If I disable the SSDT-TB3HP.aml, macOS is starting up and the eGPU just works fine.

 And regarding RP05 vs. RP21: I think it was always RP21 and I just mixed that up with my Vision D repo, where TB3 indeed sits at RP05. 

Anyway: I don't get Thunderbolt Hot plug working and with all my research I think this is because it is required to flash the TB3 controller with a modified Thunderbolt Firmware. But this requires some additional flashing tools and I don't have them. Also I don't need them, because I simply don't need hotplug for my eGPU case. I also tried to flash a modified BIOS so that we have additional Thunderbolt options for e.g. GPIO3 Force Power. But no matter what I tried, there was no additional option in the BIOS.

**Comment by fidalgodev** [comment]
> @fidalgodev Sorry for the confusion. Must have been drunk when I uploaded this file. But at least this showed me, that any TB3 SSDT variation that is addressing the correct port RP21, is crashing my MacOS as soon as the Thunderbolt eGPU case is connected. No matter if I connect it before or after booting. If I disable the SSDT-TB3HP.aml, macOS is starting up and the eGPU just works fine.
> 
> And regarding RP05 vs. RP21: I think it was always RP21 and I just mixed that up with my Vision D repo, where TB3 indeed sits at RP05.
> 
> Anyway: I don't get Thunderbolt Hot plug working and with all my research I think this is because it is required to flash the TB3 controller with a modified Thunderbolt Firmware. But this requires some additional flashing tools and I don't have them. Also I don't need them, because I simply don't need hotplug for my eGPU case. I also tried to flash a modified BIOS so that we have additional Thunderbolt options for e.g. GPIO3 Force Power. But no matter what I tried, there was no additional option in the BIOS.

Very weird indeed. So the SSDT you have "working" targets RP05, so technically it doing nothing, right?

With the correct RP21 SSDT I shared, if you turn on the computer with the eGPU already plugged in, does it work?

It's very weird that basically without any SSDT at all for thunderbolt (since it is targeting the RP05) the eGPU works ...

For me the one I shared that targets RP21 is working fine. But I stil didn't tried a real TB3 device, Only a lot of USB C stuff I have here, like external disks, and some 2.0 and 3.0 USB hubs (they do have ethernet port, and the ethernet port works on this usb c hub, but I dont think it is thunderbolt 3)

**Comment by SchmockLord** [comment]
> Very weird indeed. So the SSDT you have "working" targets RP05, so technically it doing nothing, right?

Yes, exactly. It does nothing and that is why it is working for me.

In this thread it is also said, that the hotplug via SSDT should work with most TB3 devices. But Hotplug for devices like my eGPU Case requires flashing of the TB3 firmware to a Mac specific:

https://www.tonymacx86.com/threads/gigabyte-z490-vision-d-thunderbolt-3-i5-10400-amd-rx-580.298642/

scroll down to "**Thunderbolt Bus Activation**".

> Thunderbolt with hot-plug functionality works properly just by following the build guide. However, because Apple and Intel jointly developed Thunderbolt, Apple has created a more thorough implementation compared to Windows and Linux. We can think of Thunderbolt as operating in two modes:
> Standard Mode, also known as ICM or Internal Connection Manager mode.
> Extended Mode, in which macOS implements a number of extended capabilities.
> Switching from Standard Mode to Extended Mode is purely optional, but it brings the system closer to a real Mac. Extended Mode provides these additional capabilities: 
> [...]
> eGPU support. This allows eGPUs to connect and disconnect. Additionally, a menu bar icon allows the eGPU to be disconnected safely.
> [...]

**Comment by tristan-k** [comment]
I'm on BIOS 1.40 and have several thunderbolt 2 and 3 devices to test this but with the current master branche none of them are connected.

EDIT: After updating to L1.42 Beta Bios my thunderbolt devices are recognized again.
